casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -545,105 +545,73 @@ object Validate {
-        if (validator == b.sender) {
-          // We let checkEquivocations handle this case
-          Applicative[F].pure(Left(tail))
-        } else {
-          val previousBlockJustificationHash =
-            previousLatestMessages.getOrElse(
-              validator,
-              genesis.blockHash
-            )
-          isJustificationRegression(
-            dag,
-            currentBlockJustificationHash,
-            previousBlockJustificationHash
-          ).ifM(
-            {
-              val message =
-                s"block ${PrettyPrinter.buildString(currentBlockJustificationHash)} by ${PrettyPrinter
-                  .buildString(validator)} has a lower sequence number than ${PrettyPrinter.buildString(previousBlockJustificationHash)}."
-              Log[F]
-                .warn(ignore(b, message))
-                .as(
-                  BlockStatus.justificationRegression.asLeft.asRight
-                )
-            },
-            Applicative[F].pure(Left(tail))
-          )
+      // Latest Message from sender of `b` is present in the DAG
+      case Some(curSendersBlock) =>
+        // Here we comparing view on the network by sender from the standpoint of
+        // his previous block created (current Latest Message of sender)
+        // and new block `b` (potential new Latest Message of sender)
+        val newSendersBlock = b
+        val newLMs          = ProtoUtil.toLatestMessageHashes(newSendersBlock.justifications)
+        val curLMs          = ProtoUtil.toLatestMessageHashes(curSendersBlock.justifications)
+        // We let checkEquivocations handle when sender uses old self-justification
+        val newLMsNoSelf = newLMs.filterNot(_._1 == b.sender)
+
+        // Check each Latest Message for regression (block seq num goes backwards)
+        newLMsNoSelf.toList.tailRecM {
+          // No more Latest Messages to check
+          case Nil =>
+            Applicative[F].pure(BlockStatus.valid.asRight.asRight)
+          // Check if sender of LatestMessage does justification regression
+          case newLM :: tail =>
+            val (sender, newJustificationHash) = newLM
+            val noCurLatestMessage             = !curLMs.contains(sender)
+            if (noCurLatestMessage) {
+              // If there is no LM to compare with - regression is not possible
+              Applicative[F].pure(Left(tail))
+            } else {
+              def logWarn(currentHash: BlockHash, regressiveHash: BlockHash): F[Unit] = {
+                val msg = s"block ${PrettyPrinter.buildString(regressiveHash)} by ${PrettyPrinter
+                  .buildString(sender)} has a lower sequence number than ${PrettyPrinter
+                  .buildString(currentHash)}."
+                Log[F].warn(ignore(b, msg))
+              }
+              // Compare and check for regression
+              val regressionDetected = for {
+                curJustificationHash <- curLMs(sender).pure[F]
```

Comment:
There is no need to wrap this in `F` just to be used in `for` comprehension. It can be defined as "normal" variable before `for`.

Reply:
Yeah, I do this to avoid additional bracket section, makes it less readable